File: Java/DailyLeetCode/minimumDiameterAfterMergingtrees.java
class Solution {
    int maxD;

    public int minimumDiameterAfterMerge(int[][] edges1, int[][] edges2) {
        int n1 = edges1.length + 1, n2 = edges2.length + 1;
        ArrayList<Integer>[] t1 = new ArrayList[n1];
        ArrayList<Integer>[] t2 = new ArrayList[n2];

        for (int i = 0; i < n1; i++) t1[i] = new ArrayList<>();
        for (int i = 0; i < n2; i++) t2[i] = new ArrayList<>();
        
        for (int[] e : edges1) {
            t1[e[0]].add(e[1]);
            t1[e[1]].add(e[0]);
        }
        for (int[] e : edges2) {
            t2[e[0]].add(e[1]);
            t2[e[1]].add(e[0]);
        }
        
        maxD = -1;
        boolean[] vis1 = new boolean[n1];
        calcDiam(0, t1, vis1);
        int d1 = maxD;

        maxD = -1;
        boolean[] vis2 = new boolean[n2];
        calcDiam(0, t2, vis2);
        int d2 = maxD;

        int candDiam = (d1 + 1) / 2 + (d2 + 1) / 2 + 1;
        return Math.max(candDiam, Math.max(d1, d2));
    }

    private int calcDiam(int node, ArrayList<Integer>[] t, boolean[] vis) {
        vis[node] = true;
        int maxH = -1, secMaxH = -1;

        for (int child : t[node]) {
            if (!vis[child]) {
                int h = calcDiam(child, t, vis);
                if (h > maxH) {
                    secMaxH = maxH;
                    maxH = h;
                } else if (h > secMaxH) {
                    secMaxH = h;
                }
            }
        }
        
        maxD = Math.max(maxD, maxH + secMaxH + 2);
        return maxH + 1;
    }
}

File: Python/DailyLeetCode/minimumDiameterAfterMergingTrees.py
class Solution:
    def __init__(self):
        self.maxD = -1

    def minimumDiameterAfterMerge(self, edges1, edges2):
        n1, n2 = len(edges1) + 1, len(edges2) + 1
        t1 = [[] for _ in range(n1)]
        t2 = [[] for _ in range(n2)]

        for e in edges1:
            t1[e[0]].append(e[1])
            t1[e[1]].append(e[0])
        for e in edges2:
            t2[e[0]].append(e[1])
            t2[e[1]].append(e[0])

        self.maxD = -1
        vis1 = [False] * n1
        self.calcDiam(0, t1, vis1)
        d1 = self.maxD

        self.maxD = -1
        vis2 = [False] * n2
        self.calcDiam(0, t2, vis2)
        d2 = self.maxD

        candDiam = (d1 + 1) // 2 + (d2 + 1) // 2 + 1
        return max(candDiam, max(d1, d2))

    def calcDiam(self, node, t, vis):
        vis[node] = True
        maxH, secMaxH = -1, -1

        for child in t[node]:
            if not vis[child]:
                h = self.calcDiam(child, t, vis)
                if h > maxH:
                    secMaxH, maxH = maxH, h
                elif h > secMaxH:
                    secMaxH = h

        self.maxD = max(self.maxD, maxH + secMaxH + 2)
        return maxH + 1
